File: apps/api/src/teams/teams.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  Ip,
  Param,
  Patch,
  Post,
  Put,
  Query,
} from '@nestjs/common';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import {
  AssignMemberSchema,
  CreateTeamSchema,
  UpdateTeamSchema,
  UserRole,
  type AssignMemberInput,
  type AssignMemberResult,
  type AssignableUser,
  type CreateTeamInput,
  type TeamDetail,
  type TeamRole,
  type TeamSummary,
  type UpdateTeamInput,
} from '@ihrms/shared';
import { ZodValidationPipe } from '../common/pipes/zod-validation.pipe';
import { Roles } from '../auth/decorators/scope.decorator';
import { CurrentUser } from '../auth/decorators/current-user.decorator';
import type { UserPrincipal } from '../auth/principal';
import { TeamsService } from './teams.service';

/** Team management (§2/§3.1). COMPANY_ADMIN-only; scoped to the admin's company. */
@ApiTags('teams')
@ApiBearerAuth()
@Roles(UserRole.COMPANY_ADMIN)
@Controller('teams')
export class TeamsController {
  constructor(private readonly teams: TeamsService) {}

  @Post()
  create(
    @Body(new ZodValidationPipe(CreateTeamSchema)) body: CreateTeamInput,
    @CurrentUser() actor: UserPrincipal,
    @Ip() ip: string,
  ): Promise<TeamDetail> {
    return this.teams.create(body, actor, ip);
  }

  @Get()
  list(@CurrentUser() actor: UserPrincipal): Promise<TeamSummary[]> {
    return this.teams.list(actor);
  }

  // Declared before ':id' so it is not captured by the dynamic segment.
  @Get('assignable-users')
  assignable(
    @Query('role') role: string,
    @CurrentUser() actor: UserPrincipal,
  ): Promise<AssignableUser[]> {
    return this.teams.assignableUsers(actor, role);
  }

  @Get(':id')
  get(@Param('id') id: string, @CurrentUser() actor: UserPrincipal): Promise<TeamDetail> {
    return this.teams.getDetail(id, actor);
  }

  @Patch(':id')
  update(
    @Param('id') id: string,
    @Body(new ZodValidationPipe(UpdateTeamSchema)) body: UpdateTeamInput,
    @CurrentUser() actor: UserPrincipal,
    @Ip() ip: string,
  ): Promise<TeamDetail> {
    return this.teams.update(id, body, actor, ip);
  }

  @Delete(':id')
  async remove(
    @Param('id') id: string,
    @CurrentUser() actor: UserPrincipal,
    @Ip() ip: string,
  ): Promise<{ ok: true }> {
    await this.teams.remove(id, actor, ip);
    return { ok: true };
  }

  @Put(':id/hr')
  assignHr(
    @Param('id') id: string,
    @Body(new ZodValidationPipe(AssignMemberSchema)) body: AssignMemberInput,
    @CurrentUser() actor: UserPrincipal,
    @Ip() ip: string,
  ): Promise<AssignMemberResult> {
    return this.teams.assign(id, UserRole.HR as TeamRole, body, actor, ip);
  }

  @Put(':id/manager')
  assignManager(
    @Param('id') id: string,
    @Body(new ZodValidationPipe(AssignMemberSchema)) body: AssignMemberInput,
    @CurrentUser() actor: UserPrincipal,
    @Ip() ip: string,
  ): Promise<AssignMemberResult> {
    return this.teams.assign(id, UserRole.MANAGER as TeamRole, body, actor, ip);
  }
}

File: apps/api/src/main/java/com/ihrms/teams/TeamsController.java
package com.ihrms.teams;

import com.ihrms.auth.IhrmsPrincipal;
import com.ihrms.domain.enums.UserRole;
import com.ihrms.teams.dto.TeamDtos.AssignMemberRequest;
import com.ihrms.teams.dto.TeamDtos.AssignMemberResult;
import com.ihrms.teams.dto.TeamDtos.CreateTeamRequest;
import com.ihrms.teams.dto.TeamDtos.TeamDetailView;
import com.ihrms.teams.dto.TeamDtos.TeamMemberView;
import com.ihrms.teams.dto.TeamDtos.TeamSummaryView;
import com.ihrms.teams.dto.TeamDtos.UpdateTeamRequest;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import java.util.List;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

/** Team management (contract §3.4). COMPANY_ADMIN-only, scoped to the admin's company. */
@RestController
@RequestMapping("/teams")
@PreAuthorize("hasRole('COMPANY_ADMIN')")
public class TeamsController {

  private final TeamsService teams;

  public TeamsController(TeamsService teams) {
    this.teams = teams;
  }

  @PostMapping
  @ResponseStatus(HttpStatus.CREATED)
  public TeamDetailView create(
      @Valid @RequestBody CreateTeamRequest body,
      @AuthenticationPrincipal IhrmsPrincipal.User actor,
      HttpServletRequest request) {
    return teams.create(body, actor, request.getRemoteAddr());
  }

  @GetMapping
  public List<TeamSummaryView> list(@AuthenticationPrincipal IhrmsPrincipal.User actor) {
    return teams.list(actor);
  }

  // Declared before '/{id}' so the literal path is not captured by the dynamic segment.
  @GetMapping("/assignable-users")
  public List<TeamMemberView> assignable(
      @RequestParam(name = "role", required = false) String role,
      @AuthenticationPrincipal IhrmsPrincipal.User actor) {
    return teams.assignableUsers(actor, role);
  }

  @GetMapping("/{id}")
  public TeamDetailView get(
      @PathVariable String id, @AuthenticationPrincipal IhrmsPrincipal.User actor) {
    return teams.getDetail(id, actor);
  }

  @PatchMapping("/{id}")
  public TeamDetailView update(
      @PathVariable String id,
      @Valid @RequestBody UpdateTeamRequest body,
      @AuthenticationPrincipal IhrmsPrincipal.User actor,
      HttpServletRequest request) {
    return teams.update(id, body, actor, request.getRemoteAddr());
  }

  @DeleteMapping("/{id}")
  public Map<String, Boolean> remove(
      @PathVariable String id,
      @AuthenticationPrincipal IhrmsPrincipal.User actor,
      HttpServletRequest request) {
    teams.remove(id, actor, request.getRemoteAddr());
    return Map.of("ok", true);
  }

  @PutMapping("/{id}/hr")
  public AssignMemberResult assignHr(
      @PathVariable String id,
      @RequestBody AssignMemberRequest body,
      @AuthenticationPrincipal IhrmsPrincipal.User actor,
      HttpServletRequest request) {
    return teams.assign(id, UserRole.HR, body, actor, request.getRemoteAddr());
  }

  @PutMapping("/{id}/manager")
  public AssignMemberResult assignManager(
      @PathVariable String id,
      @RequestBody AssignMemberRequest body,
      @AuthenticationPrincipal IhrmsPrincipal.User actor,
      HttpServletRequest request) {
    return teams.assign(id, UserRole.MANAGER, body, actor, request.getRemoteAddr());
  }
}
